tests/helpers/test_template.py

```diff
@@ -1162,34 +1161,90 @@ def test_as_datetime(hass: HomeAssistant, input) -> None:
     )
 
 
-def test_as_datetime_from_timestamp(hass: HomeAssistant) -> None:
-    """Test converting a UNIX timestamp to a date object."""
-    tests = [
+@pytest.mark.parametrize(
+    ("input", "output"),
+    [
         (1469119144, "2016-07-21 16:39:04+00:00"),
         (1469119144.0, "2016-07-21 16:39:04+00:00"),
         (-1, "1969-12-31 23:59:59+00:00"),
-    ]
-    for input, output in tests:
-        # expected = dt_util.parse_datetime(input)
-        if output is not None:
-            output = str(output)
+    ],
+)
+def test_as_datetime_from_timestamp(
+    hass: HomeAssistant,
+    input: int | float,
+    output: str,
+) -> None:
+    """Test converting a UNIX timestamp to a date object."""
+    assert (
+        template.Template(f"{{{{ as_datetime({input}) }}}}", hass).async_render()
+        == output
+    )
+    assert (
+        template.Template(f"{{{{ {input} | as_datetime }}}}", hass).async_render()
+        == output
+    )
+    assert (
+        template.Template(f"{{{{ as_datetime('{input}') }}}}", hass).async_render()
+        == output
+    )
+    assert (
+        template.Template(f"{{{{ '{input}' | as_datetime }}}}", hass).async_render()
+        == output
+    )
 
-        assert (
-            template.Template(f"{{{{ as_datetime({input}) }}}}", hass).async_render()
-            == output
-        )
-        assert (
-            template.Template(f"{{{{ {input} | as_datetime }}}}", hass).async_render()
-            == output
-        )
-        assert (
-            template.Template(f"{{{{ as_datetime('{input}') }}}}", hass).async_render()
-            == output
-        )
-        assert (
-            template.Template(f"{{{{ '{input}' | as_datetime }}}}", hass).async_render()
-            == output
-        )
+
+@pytest.mark.parametrize(
+    ("input", "output"),
+    [
+        (
+            "{% set dt = as_datetime('2024-01-01 16:00:00-08:00') %}",
```

Comment:
Why don't we just use the time string directly as input?

```py
{{'2024-01-01 16:00:00-08:00' | as_datetime }}
```
The docstring for `as_datetime` reads:
```py
"""Filter and to convert a time string or UNIX timestamp to datetime object."""
```

Reply:
I want to add support for `datetime.datetime` and `datetime.date` as input.
To do that, they are parsed to a string.

An annoyance of the current filter/function is that it throws an error if you use it on a datetime.datetime. Now it returns the datetime.datetime back, it avoids the need to check on the type before you use `as_datetime`

This was even worse when there was no `datetime` test yet, you had to do someting as:
```jinja
{% set dt = [now() | now().isoformat] | random %}
{% if dt is not iterable and not dt | is_number %}
  {{ as_datetime(dt) %}
{% else %}
 {{ dt }}
{% endif %}

I have updated the docstring.

Reply:
Ok. I suggest spitting this PR into three PRs, one per description bullet.

I didn't notice that we're changing three things.

Reply:
Well, the to accept datetime objects and that it will no longer error on other types as `list` and `mapping` are both the result of parsing the value to a string before using it in `dt_util.parse_datetime()`. So I can't really split that out.

I can remove the `default` parameter out of this PR, and create a separate PR for that.

Reply:
Changed the scope to only add support for the `default` parameter

To avoid breaking changes, a string input which can't be parsed to a datetime object returns `None` if no default is provided. In case a default is provided it will that for strings which can't be parsed to datetime.